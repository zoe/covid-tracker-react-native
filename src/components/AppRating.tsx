import Constants from 'expo-constants';
import { Toast, View } from 'native-base';
import React, { Component } from 'react';
import { Linking, Platform, StyleSheet, TouchableOpacity } from 'react-native';

import { colors } from '@theme';
import i18n from '@covid/locale/i18n';
import { ModalContainer } from '@covid/components/ModalContainer';
import { container, lazyInject } from '@covid/provider/services';
import { Services } from '@covid/provider/services.types';
import { IUserService } from '@covid/core/user/UserService';
import { IContentService } from '@covid/core/content/ContentService';
import { RegularBoldText, RegularText } from '@covid/components/Text';
import { isUSCountry, isSECountry } from '@covid/core/localisation/LocalisationService';

type PropsType = object;

type State = {
  isModalOpen: boolean;
  showTakeToStore: boolean;
};

const USiOSLink = `https://apps.apple.com/us/app/covid-symptom-study/id1503529611`;
const UKiOSLink = `https://apps.apple.com/gb/app/covid-symptom-study/id1503529611`;
const SEiOSLink = `https://apps.apple.com/se/app/covid-symptom-study/id1503529611`;
const AndroidLink = `market://details?id=${Constants.manifest.android.package}`;

export async function shouldAskForRating(): Promise<boolean> {
  const profile = await container.get<IUserService>(Services.User).getProfile();
  const eligibleToAskForRating = profile?.ask_for_rating ?? false;
  const askedToRateStatus = await container.get<IContentService>(Services.Content).getAskedToRateStatus();
  return !askedToRateStatus && eligibleToAskForRating;
}

export class AppRating extends Component<PropsType, State> {
  @lazyInject(Services.Content)
  private readonly contentService: IContentService;

  state = {
    isModalOpen: true,
    showTakeToStore: false,
  };

  decline = () => {
    this.contentService.setAskedToRateStatus('asked');
    this.setState({ isModalOpen: false });
  };

  declineFeedback = () => {
    this.decline();
    Toast.show({
      text: i18n.t('rating.thank-you-for-rating'),
      duration: 3000,
      position: 'top',
      textStyle: { textAlign: 'center', lineHeight: 25 },
      // style: {paddingVertical: 40}
    });
  };

  takeToStore = () => {
    this.contentService.setAskedToRateStatus('asked');
    if (Platform.OS !== 'ios') {
      Linking.openURL(AndroidLink);
    } else {
      const storeLink = isUSCountry() ? USiOSLink : isSECountry() ? SEiOSLink : UKiOSLink;
      Linking.openURL(storeLink);
    }
    this.setState({ isModalOpen: false });
  };

  askToRate = (_: any) => {
    this.setState({ showTakeToStore: true });
  };

  renderHeader = (headerText: string, subText: string) => (
    <>
      <RegularBoldText style={styles.ratingHeader}>{headerText}</RegularBoldText>
      <RegularText style={styles.ratingText}>{subText}</RegularText>
    </>
  );

  renderActionButtons = (yesLabel: string, yesAction: any, noLabel: string, noAction: any) => (
    <View style={styles.actionContainer}>
      <TouchableOpacity style={styles.ratingButton} onPress={noAction}>
        <RegularText style={styles.buttonText}>{noLabel}</RegularText>
      </TouchableOpacity>
      <View style={styles.verticalDivider} />
      <TouchableOpacity style={styles.ratingButton} onPress={yesAction}>
        <RegularText style={styles.buttonText}>{yesLabel}</RegularText>
      </TouchableOpacity>
    </View>
  );
  //

  render() {
    return (
      this.state.isModalOpen && (
        <ModalContainer>
          {this.state.showTakeToStore ? (
            <>
              {this.renderHeader(i18n.t('rating.rate-this-app'), i18n.t('rating.explanation'))}
              {this.renderActionButtons(
                i18n.t('rating.cta-rate'),
                this.takeToStore,
                i18n.t('rating.cta-later'),
                this.decline
              )}
            </>
          ) : (
            <>
              {this.renderHeader(i18n.t('rating.how-are-we-doing'), i18n.t('rating.would-you-recommend'))}
              {this.renderActionButtons(
                i18n.t('rating.cta-yes'),
                this.askToRate,
                i18n.t('rating.cta-no'),
                this.declineFeedback
              )}
            </>
          )}
        </ModalContainer>
      )
    );
  }
}

const styles = StyleSheet.create({
  verticalDivider: {
    height: '100%',
    width: 1,
    backgroundColor: colors.actionButtonBorder,
  },
  ratingText: {
    paddingBottom: 30,
    marginHorizontal: 60,
    fontSize: 14,
    textAlign: 'center',
  },
  ratingHeader: {
    paddingBottom: 10,
    fontSize: 18,
    textAlign: 'center',
  },
  actionContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    borderTopWidth: 1,
    borderColor: colors.actionButtonBorder,
  },
  buttonText: {
    textAlign: 'center',
    color: colors.linkBlue,
  },
  ratingButton: {
    width: '50%',
    height: 60,
    justifyContent: 'center',
    color: colors.linkBlue,
  },
});
