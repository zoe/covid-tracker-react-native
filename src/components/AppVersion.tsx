import Constants from "expo-constants";
import {MutedText} from "./Text";
import React from "react";


export const ApplicationVersion = () => (
  <MutedText style={{ textAlign: "center"}}>
      AppVersion: {Constants.nativeBuildVersion} - OTA: {Constants.manifest.revisionId}
  </MutedText>
);
