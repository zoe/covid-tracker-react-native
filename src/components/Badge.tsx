import { Badge } from 'native-base';
import React from 'react';
import { StyleSheet, Text } from 'react-native';

import { colors } from '@theme';

type BadgeProps = {
  children: React.ReactNode;
};

export const CoralBadge: React.FC<BadgeProps> = (props) => (
  <Badge style={styles.badge}>
    <Text style={styles.badgeText}>{props.children}</Text>
  </Badge>
);

const styles = StyleSheet.create({
  badge: {
    alignSelf: 'center',
    backgroundColor: colors.coral,
    borderRadius: 5,
    height: 30,
    marginBottom: 10,
  },
  badgeText: {
    color: colors.white,
    paddingHorizontal: 10,
  },
});
