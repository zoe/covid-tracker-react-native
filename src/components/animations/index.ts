export * from './hooks';
export { default as ProgressCircle } from './progress-circle';
export { default as Done } from './done';
