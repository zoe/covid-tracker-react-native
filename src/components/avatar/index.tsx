import { avatarBlank } from '@assets';
import * as React from 'react';
import { Image, ImageSourcePropType, ImageStyle, StyleProp, View } from 'react-native';

interface IProps {
  imgsrc?: ImageSourcePropType;
  imgStyle?: StyleProp<ImageStyle>;
  size?: number;
}

function Avatar({ imgsrc = avatarBlank, imgStyle = {}, size = 60 }: IProps) {
  return (
    <View
      style={{
        alignItems: 'center',
        backgroundColor: 'pink',
        borderColor: 'white',
        borderRadius: size,
        borderWidth: 2,
        height: size,
        justifyContent: 'center',
        overflow: 'hidden',
        width: size,
      }}
    >
      <Image
        source={imgsrc}
        style={[
          {
            aspectRatio: 1.0,
            height: undefined,
            resizeMode: 'contain',
            width: size,
          },
          imgStyle,
        ]}
      />
    </View>
  );
}

export default Avatar;
