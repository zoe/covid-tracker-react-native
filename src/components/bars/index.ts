export { default as GradientColorBar } from './gradient-color-bar';
