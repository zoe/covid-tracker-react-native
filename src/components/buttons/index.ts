export { default as ActionButton } from './ActionButton';
export { default as BackButton } from './back-button';
export { default as BigButton } from './big-button';
export { default as BrandedButton } from './branded-button';
export { default as CheckBoxButton } from './check-box-button';
export { default as IconButton } from './icon-button';
export { default as Link } from './link';
export { default as RoundIconButton } from './round-icon-button';
export { default as ShareButton } from './share-button';
export { default as ThemeButton } from './theme-button';
