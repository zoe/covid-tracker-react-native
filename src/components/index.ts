export { default as Avatar } from './avatar';
export * from './animations';
export * from './bars';
export * from './buttons';
export * from './cards';
export * from './Checkbox';
export * from './content';
export * from './DropdownField';
export { default as Fab } from './fab';
export * from './icons';
export * from './inputs';
export * from './layouts';
export * from './lists';
export * from './messaging';
export * from './profiles';
export * from './questions';
export * from './screens';
export { default as Spacer } from './spacer';
export * from './status';
export { default as Tag } from './Tag';
export * from './Text';
export { default as Triangle } from './Triangle';
export * from './typography';
export { default as Modal } from './Modal';
export { default as ModalZoe } from './ModalZoe';
export { default as DoctorProfile } from './DoctorProfile';
export { default as TextareaWithCharCount } from './TextareaWithCharCount';
