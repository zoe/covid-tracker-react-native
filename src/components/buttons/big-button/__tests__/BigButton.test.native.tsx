import { BigButton } from '@covid/components';
import * as React from 'react';
import renderer from 'react-test-renderer';

describe('Button', () => {
  it('should match snapshot', () => {
    const component = renderer.create(<BigButton onPress={() => {}}>Button Text</BigButton>);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
