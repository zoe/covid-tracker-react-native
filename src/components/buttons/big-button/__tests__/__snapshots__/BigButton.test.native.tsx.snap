// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleColor={0}
  testID="button-test-ID"
>
  <View
    accessible={true}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#565A5C",
          "borderRadius": 28,
          "borderWidth": 1,
          "fontSize": 16,
          "paddingHorizontal": 24,
          "paddingVertical": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#565A5C",
            "textAlign": "center",
          }
        }
      >
        Button Text
      </Text>
    </View>
  </View>
</RNGestureHandlerButton>
`;
